File: ios/Classes/ZegoPlatformViewFactory.m
//
//  ZegoPlatformViewFactory.m
//  Pods-Runner
//
//  Created by lizhanpeng@ZEGO on 2020/3/26.
//  Copyright © 2020 Zego. All rights reserved.
//

#import "ZegoPlatformViewFactory.h"

@interface ZegoPlatformViewFactory()

@property (nonatomic, strong) NSMutableDictionary<NSNumber*, ZegoPlatformView*> *platformViewMap;

@end

@implementation ZegoPlatformViewFactory

+ (instancetype)sharedInstance {
    static ZegoPlatformViewFactory *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[ZegoPlatformViewFactory alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _platformViewMap = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (BOOL)addPlatformView:(ZegoPlatformView *)view viewID:(NSNumber *)viewID {
    if (!viewID || !view)
        return NO;

    [self.platformViewMap setObject:view forKey:viewID];

    return YES;
}

- (BOOL)destroyPlatformView:(NSNumber *)viewID {
    if (!viewID)
        return NO;
    
    if (![[self.platformViewMap allKeys] containsObject:viewID])
        return NO;
    
    [self.platformViewMap removeObjectForKey:viewID];
    return YES;
}

- (nullable ZegoPlatformView *)getPlatformView:(NSNumber *)viewID {
    if (!viewID) {
        return nil;
    }
    
    return [self.platformViewMap objectForKey:viewID];
}

#pragma mark FlutterPlatformViewFactory Delegate

/// Called when dart invoke `createPlatformView`, that is, when Widget `UiKitView` is added to the flutter render tree
- (NSObject<FlutterPlatformView>*)createWithFrame:(CGRect)frame
                                   viewIdentifier:(int64_t)viewId
                                        arguments:(id _Nullable)args {
    
    ZegoPlatformView *view = [[ZegoPlatformView alloc] initWithRect:frame viewID:viewId];
    [self addPlatformView:view viewID:@(viewId)];
    
    return view;
}

@end

File: android/src/main/java/im/zego/zego_express_engine/ZegoPlatformViewFactory.java
//
//  android
//  im.zego.zego_express_engine
//
//  Created by Patrick Fu on 2020-04-03.
//  Copyright © 2020 Zego. All rights reserved.
//

package im.zego.zego_express_engine;

import android.content.Context;

import java.util.HashMap;

import io.flutter.plugin.common.MessageCodec;
import io.flutter.plugin.common.StandardMessageCodec;
import io.flutter.plugin.platform.PlatformView;
import io.flutter.plugin.platform.PlatformViewFactory;

public class ZegoPlatformViewFactory extends PlatformViewFactory {

    private static ZegoPlatformViewFactory instance;

    private HashMap<Integer, ZegoPlatformView> platformViews;

    private ZegoPlatformViewFactory(MessageCodec<Object> createArgsCodec) {
        super(createArgsCodec);
        this.platformViews = new HashMap<>();
    }

    static ZegoPlatformViewFactory getInstance() {
        if (instance == null) {
            synchronized (ZegoPlatformViewFactory.class) {
                if (instance == null) {
                    instance = new ZegoPlatformViewFactory(StandardMessageCodec.INSTANCE);
                }
            }
        }
        return instance;
    }

    /// Called when dart invoke `destroyPlatformView`
    void destroyPlatformView(int viewID) {
        this.platformViews.remove(viewID);
    }

    /// Get PlatfromView to pass to native when dart invoke `startPreview` or `startPlayingStream`
    ZegoPlatformView getPlatformView(int viewID) {
        return this.platformViews.get(viewID);
    }


    private void addView(int viewID, ZegoPlatformView view) {
        this.platformViews.put(viewID, view);
    }

    @Override
    public PlatformView create(Context context, int viewId, Object args) {
        ZegoPlatformView view = new ZegoPlatformView(context);
        this.addView(viewId, view);
        return view;
    }
}
